pyomo/contrib/gdpopt/ldsda.py

```diff
@@ -0,0 +1,558 @@
+        'Boolean_vars',  # list with names of the ordered Boolean variables to be reformulated
+        'UB',  # upper bound on external variable
+        'LB',  # lower bound on external variable
+    ],
+)
+
+
+@SolverFactory.register(
+    'gdpopt.ldsda',
+    doc="The LD-SDA (Logic-based Discrete-Steepest Descent Algorithm)"
+    "Generalized Disjunctive Programming (GDP) solver",
+)
+class GDP_LDSDA_Solver(_GDPoptAlgorithm):
+    """The GDPopt (Generalized Disjunctive Programming optimizer)
+    LD-SDA (Logic-based Discrete-Steepest Descent (LD-SDA) solver.
+
+    Accepts models that can include nonlinear, continuous variables and
+    constraints, as well as logical conditions.
+    """
+
+    CONFIG = _GDPoptAlgorithm.CONFIG()
+    _add_mip_solver_configs(CONFIG)
+    _add_nlp_solver_configs(CONFIG, default_solver='ipopt')
+    _add_nlp_solve_configs(
+        CONFIG, default_nlp_init_method=restore_vars_to_original_values
+    )
+    _add_tolerance_configs(CONFIG)
+    _add_ldsda_configs(CONFIG)
+
+    algorithm = 'LDSDA'
+
+    # Override solve() to customize the docstring for this solver
+    @document_kwargs_from_configdict(CONFIG, doc=_GDPoptAlgorithm.solve.__doc__)
+    def solve(self, model, **kwds):
+        return super().solve(model, **kwds)
+
+    def _log_citation(self, config):
+        config.logger.info(
+            "\n"
+            + """- LDSDA algorithm:
+        Bernal DE, Ovalle D, Liñán DA, Ricardez-Sandoval LA, Gómez JM, Grossmann IE.
+        Process Superstructure Optimization through Discrete Steepest Descent Optimization: a GDP Analysis and Applications in Process Intensification.
+        Computer Aided Chemical Engineering 2022 Jan 1 (Vol. 49, pp. 1279-1284). Elsevier.
+        https://doi.org/10.1016/B978-0-323-85159-6.50213-X
+        """.strip()
+        )
+
+    def _solve_gdp(self, model, config):
+        """Solve the GDP model.
+
+        Parameters
+        ----------
+        model : ConcreteModel
+            The GDP model to be solved
+        config : ConfigBlock
+            GDPopt configuration block
+        """
+        logger = config.logger
+        self.log_formatter = (
+            '{:>9}   {:>15}   {:>20}   {:>11.5f}   {:>11.5f}   {:>8.2%}   {:>7.2f}  {}'
+        )
+        self.best_direction = None
+        self.current_point = tuple(config.starting_point)
+        self.explored_point_set = set()
+
+        # Create utility block on the original model so that we will be able to
+        # copy solutions between
+        util_block = self.original_util_block = add_util_block(model)
+        add_disjunct_list(util_block)
+        add_algebraic_variable_list(util_block)
+        add_boolean_variable_lists(util_block)
+
+        # We will use the working_model to perform the LDSDA search.
+        self.working_model = model.clone()
+        # TODO: I don't like the name way, try something else?
+        self.working_model_util_block = self.working_model.component(util_block.name)
```

Comment:
You may be able to use `self.working_model.find_component(util_block)` to address this TODO. I don't remember how block scoping works with `find_component` but you could try it.

Reply:
Resolved. Thank you for your suggestion.